# Worm Perturb-seq (WPS)
This repository provides a data processing pipeline of Worm Perturb-Seq (WPS) technology.

------------------------------------------------------------------------
Please note that we are still actively developing this pipeline. Please contact us if you see any issues!
------------------------------------------------------------------------

**WPS** is an end-to-end massively parallel RNAi and RNA-seq technology in model animal C. elegans. WPS involves both an experimental and computational pipeline that seamlessly cover the process from culturing animals to identify differentially expressed (DE) genes in each perturbation. The workflow is summarized in the following figure:

<img src="man/figures/GitHub.png" width="600"/>

This repository details the procedures to process WPS data, starting from the fastq files generated by NGS sequencing to the production of DE genes in each condition. 

To find out more details about **WPS**, please read our manuscript:

[Title and authors](https://bioRxiv_link)


# Table of contents
1. [Dependencies](#dependencies-)
2. [Quick Start](#quick-start)
3. [Contact](#contact)
4. [Related Manuscripts](#related-manuscripts)


## Dependencies<a name="dependencies"></a>

WPS data processing involves multiple tools and considerable computational power. We recommend setting up a local server or computational cluster for processing the data. 

* Via Foundry: the raw data processing uses a pre-built dolphinNext pipeline for the ease of use and for easy management of large-scale data. Please register at [Via Foundry](https://viafoundry.umassmed.edu) before you start. 
If you intend to process the raw data (e.g., aligning the reads to genome) manually, please see [1_process_the_raw_data](step1_process_raw_data/) for more information (but you may need to install some dependencies that are not included in this list). We strongly recommend using the dolphinNext pipeline for the sake of reproducibility and robustness.
* Python (what version?)
* R (what verison?)

 


``` python

```

``` r
if (!require("devtools", quietly = TRUE))
    install.packages("devtools")
devtools::install_github("XuhangLi/wpsDE")
```

## Quick Start<a name="quick-start"></a>

The following code is a quick example of running the full **wpsDE**. The function `WPS_DE()` takes in a count table that is a gene-by-sample dataframe of read counts of the input dataset and a metadata table that is a dataframe with required metadata (such as covariates) to run DE analysis. For more details, please check on the help page by `?WPS_DE`. For a test run:

``` r
data("countTable")
data("metaDataTable")
result <- WPS_DE(countTable, metaDataTable)
```

The required metadata of `WPS_DE()` includes the following columns:

- `sampleID`: unique IDs for each sample that correspond to the column names in countTable.
- `covTreatment`: covariate indicating experimental treatments to be tested for (e.g., RNAi conditions). Must have a 'control' condition if control-dependent DE analysis is deisred.
- `covBatch`: covariate indicating experimental batches, which by default is the replicate batch. Other reasonable batch labels within each library may also be used. 
- `libID`: unique IDs for identifying the sequencing library of each sample. WPS DE analysis is first conducted at individual sequencing library level thus this ID is used to match samples pooled in the same library.

The output of `WPS_DE()` is a list tables for the DE results of each condition. Key columns include: 

-   `log2FoldChange_raw`: log2FoldChange from DESeq2. Adding the suffix '_raw' is to emphasize this metric is directly from DESeq2 model without applying any shrinkage algorithm.
-   `DE_source`: The DE result for this gene is based on which type of DE analysis, either control-dependent (vs. control) or independent (vs. control-independent null).
-   `empirical_pvalue`: Empirical p-values based on corrected test statistic. This is the final DE testing p-value of WPS DE framework.
-   `FDR`: False Discovery Rate (FDR) of the DE test. This is the final DE testing FDR of WPS DE framework.

It is notable that running full **wpsDE** analysis on a real data set (e.g., 10,000 genes x 100 conditions x 3 replicates) can be time consuming. It is expected to see prolonged run time up to several hours. We are working on integrating parallelization functions to increase the performance.


For users interested in custom use of functionalities in **wpsDE**, the following code gives a quick example on two common applications:


_To analyze the empirical null of a test statistic matrix, run_
``` r
data(example_stat_table)
result <- fit_empirical_null(example_stat_table)
```
The test statistic matrix can by obtained from a user-defined DESeq2 analysis of their own data (using the `stat` column from DESeq2 output). The `result` object includes the empirical p-value matrix calculated from the input test statistic matrix. 



_To perform control-independent DE analysis, run_
``` r
data(example_dds)
adaZmat <- fit_main_population(example_dds)
result <- control_independent_DE(example_dds, adaZmat)
```
The input `example_dds` is a DESeqDataSet objective that contains the input data. The DESeqDataSet should include specific covariate names, see details on the example data (`?example_dds`) for instructions. The returning object is a list of DE results for each conditions in the input DESeqDataSet.


## Contact<a name="contact"></a>

Any questions or suggestions on `wpsDE` are welcomed! Please report it on [issues](https://github.com/XuhangLi/wpsDE/issues), or contact Xuhang Li ([xuhang.li\@umassmed.edu](mailto:xuhang.li@umassmed.edu){.email}).

## Related Manuscripts<a name="related-manuscripts"></a>
-  **wpsDE** integrates an algorithm (adaTiss) for robust fitting of the expression levels. See original paper at: 
    -   **adaTiss**: [Wang M, Jiang L, Snyder MP. AdaTiSS: a novel data-Adaptive robust method for identifying Tissue Specificity Scores. <em>Bioinformatics</em> (2021).]([https://www.nature.com/articles/s41587-023-01772-1](https://academic.oup.com/bioinformatics/article/37/23/4469/6306407))

## Acknowledgement
We thank Dongyuan Song for his help in building this package.


