# Worm Perturb-seq (WPS)
This repository provides a data processing pipeline of Worm Perturb-Seq (WPS) technology.

------------------------------------------------------------------------
Please note that we are still actively developing this pipeline. Please contact us if you see any issues!
------------------------------------------------------------------------

**WPS** is an end-to-end massively parallel RNAi and RNA-seq technology in model animal C. elegans. WPS involves both an experimental and computational pipeline that seamlessly cover the process from culturing animals to identify differentially expressed (DE) genes in each perturbation. The workflow is summarized in the following figure:

<img src="man/figures/GitHub.png" width="600"/>

This repository details the procedures to process WPS data, starting from the fastq files generated by NGS sequencing to the production of DE genes in each condition. 

To find out more details about **WPS**, please read our manuscript:

[Title and authors](https://bioRxiv_link)


# Table of contents
1. [Dependencies](#dependencies-)
2. [Walkthrough](#walkthrough)
3. [Contact](#contact)
4. [Related Manuscripts](#related-manuscripts)


## Dependencies<a name="dependencies"></a>

WPS data processing involves multiple tools and considerable computational power. We recommend setting up a local server or computational cluster for processing the data. 

* Via Foundry: the raw data processing uses a pre-built dolphinNext pipeline for the ease of use and for easy management of large-scale data. Please register at [Via Foundry](https://viafoundry.umassmed.edu) before you start. 
If you intend to process the raw data (e.g., aligning the reads to genome) manually, please see [1_process_the_raw_data](step1_process_raw_data/) for more information (but you may need to install some dependencies that are not included in this list). We strongly recommend using the dolphinNext pipeline for the sake of reproducibility and robustness.
* Python (what version?)
* R (what verison?)

## Walkthrough<a name="walkthrough"></a>

The pipeline includes three major steps: [step1_process_raw_dataalgorithm](1_iMAT++), [original iMAT algorithm](2_iMAT), [Flux Potential Analysis (FPA)](3_FPA), [metabolic distance calculator](MetabolicDistance), and [categorizer of gene expression data](CatExp). Please see the instruction within each module for running a test.

The followings are descriptions on each module (folder) listed.

[1_iMAT++](1_iMAT++): The iMAT++ module. This folder contains all the input, scripts and example codes for running the iMAT++ for <i>C. elegans</i> tissue-level integration and application on other models/datasets. 

[2_iMAT](2_iMAT): The original iMAT module. This folder contains all the input, scripts and example codes for running the original iMAT for <i>C. elegans</i> tissue-level integration. 

[3_FPA](3_FPA): The Flux Potential Analysis (FPA) module. This folder contains all the input, scripts and example codes for running the FPA for <i>C. elegans</i> tissue-level integration and application on other models/datasets. 

[MetabolicDistance](MetabolicDistance): The metabolic distance calculator module. This folder contains all the input, scripts and example codes for calculating the metabolic distance for any metabolic network model. The output distance matrix is required to run FPA. The output for the <i>C. elegans</i> models used in the reference study are available in the pertaining folders. 

[CatExp](CatExp): CatExp.py provides a set of functions for systematic categorization of gene expression levels. This folder also contains example inputs (gene expression datasets) and outputs (tables of categorized genes) used in the walkthrough help guide. The examples include the <i>C. elegans</i> tissue-level dataset used in the paper. 

[bins](bins): The shared functions for running the above mentioned analysis. These functions include modified version of some COBRA Toolbox functions and common new functions such as a molecular weight calculator.

[input](input): The shared input for running the above mentioned analysis. These inputs include <i>C. elegans</i> metabolic model and other input information.


## Contact<a name="contact"></a>

Any questions or suggestions on `wpsDE` are welcomed! Please report it on [issues](https://github.com/XuhangLi/wpsDE/issues), or contact Xuhang Li ([xuhang.li\@umassmed.edu](mailto:xuhang.li@umassmed.edu){.email}).

## Related Manuscripts<a name="related-manuscripts"></a>
-  **wpsDE** integrates an algorithm (adaTiss) for robust fitting of the expression levels. See original paper at: 
    -   **adaTiss**: [Wang M, Jiang L, Snyder MP. AdaTiSS: a novel data-Adaptive robust method for identifying Tissue Specificity Scores. <em>Bioinformatics</em> (2021).]([https://www.nature.com/articles/s41587-023-01772-1](https://academic.oup.com/bioinformatics/article/37/23/4469/6306407))

## Acknowledgement
We thank Dongyuan Song for his help in building this package.


